lib/puppet/http/service/compiler.rb

```diff
@@ -129,6 +129,71 @@ def post_catalog(name, facts:, environment:, configured_environment: nil, transa
     [response, deserialize(response, Puppet::Resource::Catalog)]
   end
 
+  #
+  # @api private
+  #
+  # Submit a POST request to request a catalog to the server using v4 endpoint
+  #
+  # @param [String] certname The name of the node for which to compile the catalog.
+  # @param [Hash] persistent A hash containing two required keys, facts and catalog,
+  #   which when set to true will cause the facts and reports to be stored in
+  #   PuppetDB, or discarded if set to false.
+  # @param [String] environment The name of the environment for which to compile the catalog.
+  # @param [Hash] facts A hash with a required values key, containing a hash of all the
+  #    facts for the node. If not provided, Puppet will attempt to fetch facts for the node
+  #    from PuppetDB.
+  # @param [Hash] trusted_facts A hash with a required values key containing a hash of
+  #    the trusted facts for a node
+  # @param [String] transaction_uuid The id for tracking the catalog compilation and
+  #    report submission.
+  # @param [String] job_id The id of the orchestrator job that triggered this run.
+  # @param [Hash] options A hash of options beyond direct input to catalogs. Options:
+  #    - prefer_requested_environment Whether to always override a node's classified 
+  #      environment with the one supplied in the request. If this is true and no environment
+  #      is supplied, fall back to the classified environment, or finally, 'production'.
+  #    - capture_logs Whether to return the errors and warnings that occurred during
+  #      compilation alongside the catalog in the response body.
+  #    - log_level The logging level to use during the compile when capture_logs is true.
+  #      Options are 'err', 'warning', 'info', and 'debug'.
+  #
+  # @return [Array<Puppet::HTTP::Response, Puppet::Resource::Catalog, Array<String>>] An array
+  #   containing the request response, the deserialized catalog returned by
+  #   the server and array containing logs (log array will be empty if capture_logs is false)
+  #
+  def post_catalog4(certname, persistence, environment, facts: nil, trusted_facts: nil, transaction_uuid: nil, job_id: nil, options: nil)
+    body = {
+      certname: certname,
+      persistence: persistence,
+      environment: environment,
+      transaction_uuid: transaction_uuid,
+      job_id: job_id,
+      options: options
+    }
+    body[:facts] = { values: facts } unless facts.nil?
```

Comment:
The docs say the `facts` should be a `Hash`, but I think the tests are passing a `Puppet::Node::Facts` object. Is the latter expected/required? Maybe call `facts.to_json` here so it's clear we're converting from Facts to Hash here?

Reply:
Yeah i think this is for sure something we need to figure out. Looks like the representation in the current test case would not actually do what we want:
```
{"certname"=>"ziggy",
 "persistence"=>{"facts"=>true, "catalog"=>true},
 "environment"=>"testing",
 "transaction_uuid"=>"ec3d2844-b236-4287-b0ad-632fbb4d1ff0",
 "job_id"=>"1",
 "options"=>{"prefer_requested_environment"=>false, "capture_logs"=>false},
 "facts"=>
  {"values"=>
    {"name"=>"ziggy",
     "values"=>{},
     "timestamp"=>"2021-04-14T16:24:05.049041356-07:00"}},
 "trusted_facts"=>{"values"=>{}}}

```

From what i can tell we expect this to literally be a hash like `"{"values": {"my": "facts"}}`. I will double check with Froyo about this. thanks!

Reply:
Confirmed with Froyo that facts must be represented as structured data for this endpoint. Added some validation and updated tests.